### PR TITLE
Remove vhadd from ARM Codegen patterns for (a+b)/2 because of non-Halide wrapping.

### PR DESCRIPTION
**Pull request by aam**
ARM ```vhadd``` instruction (a+b)/2 doesn't comply with Halide semantic of wrapping add operands
therefore this pull request removes vhadd patterns from ARM Codegen.

test/arm-test added with this PR demonstrates the problem(executed on Nexus 6 device). Snippets from ```adb log``` output are below.

Before
```
D TestActivity: Starting the test...
E halide_native: @0 (246+246)/2 expected to be 118, but turned out to be 246
E halide_native: @1 (247+247)/2 expected to be 119, but turned out to be 247
E halide_native: @2 (248+248)/2 expected to be 120, but turned out to be 248
E halide_native: @3 (249+249)/2 expected to be 121, but turned out to be 249
E halide_native: @4 (250+250)/2 expected to be 122, but turned out to be 250
E halide_native: @5 (251+251)/2 expected to be 123, but turned out to be 251
E halide_native: @6 (252+252)/2 expected to be 124, but turned out to be 252
E halide_native: @7 (253+253)/2 expected to be 125, but turned out to be 253
E halide_native: @8 (254+254)/2 expected to be 126, but turned out to be 254
E halide_native: @9 (255+255)/2 expected to be 127, but turned out to be 255
I TestActivity: 8-bit Failed
E halide_native: @0 expected 32758 didn't match actual 65526
E halide_native: @1 expected 32759 didn't match actual 65527
E halide_native: @2 expected 32760 didn't match actual 65528
E halide_native: @3 expected 32761 didn't match actual 65529
E halide_native: @4 expected 32762 didn't match actual 65530
E halide_native: @5 expected 32763 didn't match actual 65531
E halide_native: @6 expected 32764 didn't match actual 65532
E halide_native: @7 expected 32765 didn't match actual 65533
E halide_native: @8 expected 32766 didn't match actual 65534
E halide_native: @9 expected 32767 didn't match actual 65535
I TestActivity: 16-bit Failed
E halide_native: @0 expected 2147483638 didn't match actual 4294967286
E halide_native: @1 expected 2147483639 didn't match actual 4294967287
E halide_native: @2 expected 2147483640 didn't match actual 4294967288
E halide_native: @3 expected 2147483641 didn't match actual 4294967289
E halide_native: @4 expected 2147483642 didn't match actual 4294967290
E halide_native: @5 expected 2147483643 didn't match actual 4294967291
E halide_native: @6 expected 2147483644 didn't match actual 4294967292
E halide_native: @7 expected 2147483645 didn't match actual 4294967293
E halide_native: @8 expected 2147483646 didn't match actual 4294967294
E halide_native: @9 expected 2147483647 didn't match actual 4294967295
I TestActivity: 32-bit Failed
```

After
```
D TestActivity: Starting the test...
I TestActivity: 8-bit Pass
I TestActivity: 16-bit Pass
I TestActivity: 32-bit Pass
```

### CODE REVIEW

**Comment by abadams** [comment]
My bad. Those patterns were from before we had a strict notion of how Halide's type system worked.

Regarding the testing - the full set of correctness tests run every night on arm (on a chromebook). Maybe it would enough to add an overflow-behavior-checking correctness test? Then again, that's hard for others to run. Do people think it's reasonable to just make it one of the correctness tests?

**Comment by aam** [comment]
Ah, I didn't know test_correctness runs on arm regularly. If that's the case adding simple overflow test to test_correctness should be fine.

**Comment by steven-johnson** [comment]
Might be interesting to try to set up a QEMU based ARM emulator test
environment... obviously this wouldn't inform performance issues at all,
but would make it a lot easier for people-other-than-Andrew to do
correctness testing for ARM.

File: src/CodeGen_ARM.cpp [hunk 1]
```diff
@@ -327,22 +327,6 @@ CodeGen_ARM::CodeGen_ARM(Target t) : CodeGen_Posix(t) {
...
-        averagings.push_back(Pattern("vhadds.v2i32", 2, (wild_i32x2 + wild_i32x2)));
-        averagings.push_back(Pattern("vhaddu.v2i32", 2, (wild_u32x2 + wild_u32x2)));
-
-        // 128-bit
-        averagings.push_back(Pattern("vhadds.v16i8", 16, (wild_i8x_  + wild_i8x_)));
-        averagings.push_back(Pattern("vhaddu.v16i8", 16, (wild_u8x_  + wild_u8x_)));
-        averagings.push_back(Pattern("vhadds.v8i16",  8, (wild_i16x_ + wild_i16x_)));
-        averagings.push_back(Pattern("vhaddu.v8i16",  8, (wild_u16x_ + wild_u16x_)));
-        averagings.push_back(Pattern("vhadds.v4i32",  4, (wild_i32x_ + wild_i32x_)));
-        averagings.push_back(Pattern("vhaddu.v4i32",  4, (wild_u32x_ + wild_u32x_)));
```
Comment (dsharletg) [inline src/CodeGen_ARM.cpp]: Two comments on this: 

First, I think we should keep the i32 variants of this. Overflow for i32 is undefined behavior in Halide, so it should be fine.

Second, we need to find a way to continue to generate these instructions. These instructions seem very important for optimization. Maybe the pattern needs to be modified such that the two operands are upcasted to a wider type before being downcast back down?

Comment (abadams) [inline src/CodeGen_ARM.cpp]: The patterns that use casts are included as well. They're generated
programmatically in the for loop above the deleted code. Note that
test_simd_op_check needs to be updated too once these patterns are removed.

**Comment by dsharletg** [comment]
I just realized, it is suspicious that the fuzz tester didn't see this. Maybe we should bias the fuzz tester towards producing particular integers more likely to be relevant for patterns like this (such as 2)?

**Comment by abadams** [comment]
@srj - do you think the testing component of this is something the platforms test could cover on Android? If so I'd like to accept this PR except for the arm-test part.

I would like everything under test/ to be runnable by a script every night, and that isn't the case here.

**Comment by aam** [comment]
I think @abadams meant to ask @steven-johnson

**Comment by abadams** [comment]
Oops. Sorry srj

**Comment by steven-johnson** [comment]
On Thu, May 7, 2015 at 1:15 PM, Andrew Adams <notifications@github.com>
wrote:

> @srj <https://github.com/srj> - do you think the testing component of
> this is something the platforms test could cover on Android? If so I'd like
> to accept this PR except for the arm-test part.
>

test/platforms isn't (yet) intended for automation (hence my uncertainty
about name and location), so, no, not soon.

**Comment by aam** [comment]
I would be happy to move test/arm-test to test-correctness, but don't have arm device where I can run Halide on to confirm that it works.
So was looking at using qemu-arm, but without good results yet neither.

**Comment by abadams** [comment]
A working Halide-testing qemu setup would be nice. I'll investigate too

**Comment by dsharletg** [comment]
Aside from the testing issue, this seems like a good change. We're still testing the existing correct patterns for this instruction in simd_op_check the same as we always were. I suggest we remove the extra testing stuff from this PR and save it for another, and then merge this one.

**Comment by aam** [comment]
Sounds good @dsharletg . Took out arm-test in ab5af6e.

**Pull request merged**